File: 0752-open-the-lock/0752-open-the-lock.java
class Solution {
    public void fillNeighbors(Queue<String> que, StringBuilder curr, Set<String> dead) {
        for (int i = 0; i < 4; i++) {
            char ch = curr.charAt(i);

            char dec = ch == '0' ? '9' : (char) (ch - 1);
            char inc = ch == '9' ? '0' : (char) (ch + 1);

            curr.setCharAt(i, dec);
            String decStr = curr.toString();
            if (!dead.contains(decStr)) {
                dead.add(decStr);
                que.add(decStr);
            }

            curr.setCharAt(i, inc);
            String incStr = curr.toString();
            if (!dead.contains(incStr)) {
                dead.add(incStr);
                que.add(incStr);
            }
            curr.setCharAt(i, ch);
        }
    }

    public int openLock(String[] deadends, String target)  {
        Set<String> dead = new HashSet<>(Arrays.asList(deadends));

        StringBuilder start = new StringBuilder("0000");
        if (dead.contains(start.toString()))
            return -1;

        Queue<String> que = new LinkedList<>();
        que.add(start.toString());

        int level = 0;
        while (!que.isEmpty()) {
            int n = que.size();

            while (n-- > 0) {
                String curr = que.poll();
                assert curr != null;
                if (curr.equals(target))
                    return level;

                fillNeighbors(que, new StringBuilder(curr), dead);
            }
            level++;
        }
        return -1;
    }
}

File: LeetCode/0752-open-the-lock/0752-open-the-lock.py
from collections import deque

class Solution:
    def fill_neighbors(self, que, curr, dead):
        for i in range(4):
            ch = curr[i]

            dec = '9' if ch == '0' else chr(ord(ch) - 1)
            inc = '0' if ch == '9' else chr(ord(ch) + 1)

            curr[i] = dec
            if ''.join(curr) not in dead:
                dead.add(''.join(curr))
                que.append(list(curr))

            curr[i] = inc
            if ''.join(curr) not in dead:
                dead.add(''.join(curr))
                que.append(list(curr))
            curr[i] = ch

    def openLock(self, deadends: List[str], target: str) -> int:
        dead = set(deadends)

        start = ['0', '0', '0', '0']
        if ''.join(start) in dead:
            return -1

        que = deque()
        que.append(start)

        level = 0
        while que:
            n = len(que)

            for _ in range(n):
                curr = que.popleft()
                if curr == list(target):
                    return level

                self.fill_neighbors(que, curr, dead)
            level += 1
        return -1